functional_tests/steps/bundles.py

```diff
@@ -121,3 +129,337 @@ def the_user_can_see_the_bundle_details_with_creator(context: Context) -> None:
+            bundle.save()
+
+            if bool(bundle_dets["preview_teams"]) and hasattr(context, "team"):
+                BundleTeam.objects.create(parent=bundle, team=context.team)
+                context.bundleteams.append({"parent": bundle, "team": context.team})
+            bundle.save()
+
+            if bool(bundle_dets["add_rel_cal"]) and hasattr(context, "release_calendar_page"):
+                page = context.release_calendar_page
+
+                bundle.release_calendar_page = context.release_calendar_page
+                BundlePage.objects.create(parent=bundle, page=page)
+                context.bundlepages.append(BundlePageFactory(parent=bundle, page=page))
+            else:
+                now = (datetime.now() + timedelta(hours=4)).strftime("%Y-%m-%d %H:%M")
+                bundle.scheduled_date = now
+            bundle.save()
+
+            if bool(bundle_dets["add_stat_page"]) and hasattr(context, "statistical_article_page"):
+                page = context.statistical_article_page
+                BundlePage.objects.create(parent=bundle, page=page)
+                context.bundlepages.append(BundlePageFactory(parent=bundle, page=page))
+            bundle.save()
+            context.bundles.append(bundle)
+
+        sorted(context.bundles, key=lambda x: x.name)
+
+
+# Bundles UI Triggers
+@when("the {user_role} logs in")
+def log_in_user_by_role(context: Context, user_role: str) -> None:
+    context.page.goto(f"{context.base_url}/admin/login/")
+    context.page.get_by_placeholder("Enter your username").fill(context.users[user_role]["username"])
+    context.page.get_by_placeholder("Enter password").fill(context.users[user_role]["password"])
+    context.page.get_by_role("button", name="Sign in").click()
+
+
+# Bundles UI Consequences
+@then("the logged in user can create a bundle")
+def add_bundle_details(context: Context) -> None:
+    bundle_name = "Bundle UI Test Bundle"
+    context.page.goto(context.base_url + "/admin/")
+    expect(context.page.get_by_role("link", name="Add bundle")).to_be_visible()
+    context.page.get_by_role("link", name="Add bundle").click()
+    context.page.get_by_role("textbox", name="Name*").fill(bundle_name)
+    context.page.get_by_role("button", name="Save as draft").click()
+    expect(context.page.get_by_text("Bundle successfully created.")).to_be_visible()
+
+
+@then("the logged in user fails to create a bundle due to mandatory field")
+def fail_to_save_bundle_name(context: Context) -> None:
+    context.page.goto(context.base_url + "/admin/")
+    expect(context.page.get_by_role("link", name="Add bundle")).to_be_visible()
+    context.page.get_by_role("link", name="Add bundle").click()
+    context.page.get_by_role("button", name="Save as draft").click()
+    expect(context.page.get_by_role("status")).to_contain_text("The bundle could not be created due to errors.")
+    expect(context.page.locator("#panel-child-name-errors")).to_contain_text("This field is required.")
```

Comment:
Ideally we would split this to two steps
* the logged in user saves a bundle
* the logged in user sees a validation error

This allows you to avoid repetition, and better structures your test with steps that do something and assertions that check something.